docs/technical-design/resovler-design.md

```diff
@@ -0,0 +1,146 @@
+# Resolver design
+
+## Background
+Resolver functions are responsible for handling requests made to the MC-Review GraphQL API and populating data for the response. Between receiving the request data and returning a response, the resolvers perform various tasks for generating the response.
+
+These tasks include:
+- Business logic, data transformation, and validations
+- Database operations
+- Third party API calls
+   - SES Email
+   - LaunchDarkly Feature Flags
+- Error handling
+
+The resolvers often do not directly contain the code to perform these tasks, instead they call functions specific for each task. Importantly we use types defined in our [domain models](design-patterns.md#domain-models) as the internal communication interface for different services as an attempt to decrease the coupling of our code.
+
+Errors from these tasks propagate up to the resolver where it will be handled and a response sent to the client. See docs about [error handling](error-handling.md) for details.
+
+## Diagram
```

Comment:
I think this diagram could be made simpler. You are getting at the heart of things but it might be too complex for mermaid

My notes:
* With the arrows coming in at the center of the box it's unclear that the top level resolvers are represented by the big box on the left. 
* I think it would be cool to show the individual modules as being in separate boxes with the dependencies they wrap, rather than all of our wrappers being in a box together
* Right now, our handlers are an exception where we are directly not wrapping the interface with Apollo Server, but that is another dependency we have that probably should be called out. The inputs / return values (including the thrown exceptions) from our handlers are in Apollo Server's language.
* maybe call the "hppToDomainType" stuff our "proto" module because those functions are intentionally boxed up in there to wrap that dependency. 

We can do some sketching together if you like

Reply:
I agree with this in terms of maybe too complex. Def recommend taking more time here or working with others. This is important but also challenging task!

It if is difficult to know how to adjust the diagram -- I  recommend running by each member of the team (especially @pearl-truss as the newest team member and least indoctrinated into how we all talk about the code) and ask them what they notice or what they take way from it.

Reply:
> might be too complex for mermaid

This diagram was a pain and was too complex for a mermaid. I'm thinking of just sketching this on the Miro board, adding a screenshot and linking to the docs.

> I recommend running by each member of the team (especially @pearl-truss as the newest team member and least indoctrinated into how we all talk about the code) and ask them what they notice or what they take way from it.

This is a great idea. I will fix the diagram, and maybe @pearl-truss can give me feedback either here or a Zoom meeting when she is back from OOO.

Reply:
> Right now, our handlers are an exception where we are directly not wrapping the interface with Apollo Server, but that is another dependency we have that probably should be called out. The inputs / return values (including the thrown exceptions) from our handlers are in Apollo Server's language.

Could you explain this more?

Reply:
> Could you explain this more?

So the general rule that this diagram is demonstrating is that all of our 3rd party code dependencies are getting wrapped in a module of our own design. That way we control the interface (using domain models) and contain the dependency's ways of doing things to the one module. The one exception to that is Apollo Server. That's a 3rd party dependency that we have not wrapped up in our own module. We have instead implemented our core business logic directly in Apollo's handlers. So, for instance, that's the only place in our code where we are throwing errors instead of returning them because that is how Apollo handles errors. 

I'm not sure exactly how to get that into this diagram but since this is showing off all our dependencies I think it would be good to call this out as well

Reply:
Lets pair tomorrow at our 1:1 on this?